src/loaders/genome_collection/parse_computed_genome_attribs.py

```diff
@@ -191,6 +234,233 @@ def checkm2(root_dir, kbase_collection, load_ver):
+        }
+    else:
+        # check if the trait information is consistent
+        existing_trait = global_traits_meta[trait_id]
+        if (existing_trait[_SYS_TRAIT_NAME] != trait_name or
+                existing_trait[_SYS_TRAIT_DESCRIPTION] != trait_description or
+                existing_trait[_SYS_TRAIT_CATEGORY] != trait_category):
+            raise ValueError(f'Inconsistent trait information for trait {trait_id}')
+
+
+def _parse_categories(traits_meta: Dict[str, Dict[str, str]]) -> List[ColumnCategory]:
+    # Parse the trait categories from the global traits dictionary
+
+    categories = {}
+    # loop over each trait in traits_meta
+    for trait in traits_meta.values():
+        trait_category = trait[_SYS_TRAIT_CATEGORY]
+
+        # if the trait_category is not available, create a new ColumnCategory object with empty columns
+        if trait_category not in categories:
+            categories[trait_category] = ColumnCategory(category=trait_category, columns=[])
+
+        # add the trait to the appropriate category's list of columns
+        categories[trait_category].columns.append(ColumnInformation(id=str(trait[_SYS_TRAIT_INDEX]),
+                                                                    name=trait[_SYS_TRAIT_NAME],
+                                                                    description=trait[_SYS_TRAIT_DESCRIPTION],
+                                                                    type=ColumnType.COUNT.value))
+
+    return list(categories.values())
+
+
+def _is_float_int(num: Any) -> bool:
+    # Check if a number is an integer or a float with an integer value
+    return isinstance(num, int) or (isinstance(num, float) and num.is_integer())
+
+
+def _append_cell(
+        heatmap_row: HeatMapRow,
+        trait_idx: int,
+        cell_count: int,
+        min_value: float | int,
+        max_value: float | int,
+        trait_val: float | int | bool = 0) -> (float | int, float | int):
+    # Append a cell to the heatmap row and return the global min and max values
+
+    cell = Cell(celid=str(cell_count), colid=str(trait_idx), val=trait_val)
+    heatmap_row.cells.append(cell)
+
+    return min(min_value, trait_val), max(max_value, trait_val)
+
+
+def _parse_rows(
+        traits_meta: Dict[str, Dict[str, Union[int, str]]],
+        traits_val: Dict[str, List[Dict[str, Union[int, float, bool]]]],
+        ensure_ints: bool = False) -> (List[HeatMapRow], Union[float, int], Union[float, int]):
+    min_value, max_value, cell_count = float('inf'), float('-inf'), 0
+    heatmap_rows, trait_idxs = [], set([trait[_SYS_TRAIT_INDEX] for trait in traits_meta.values()])
+    for data_id, traits_val_list in traits_val.items():
+        heatmap_row = HeatMapRow(kbase_id=data_id, cells=[])
+        visited_traits = set()
+        for trait_val_info in traits_val_list:
+            trait_idx, trait_val = trait_val_info[_SYS_TRAIT_INDEX], trait_val_info[_SYS_TRAIT_VALUE]
```

Comment:
Shouln't trait_val be checked that it matched the column type and be coerced to that type if necessary?

Reply:
it was done by `if ensure_ints`. I currently take every columns as count type and ensure int is applied to every cell.

Reply:
:+1: